### ISSUE TITLE
(restricted user issue) ModuleNotFoundError: No module named 'comtypes.gen.stdole'

### ISSUE THREAD

**Issue by vasily-v-ryabov** [enthought/comtypes#161]
Creating wrapper is failing for `UIAutomationCore.dll` if I run Python 3.6.5 32-bit as restricted user (installation was done as Administrator of course). I'm planning to debug this issue when I have time.
```python
    self.UIA_dll = comtypes.client.GetModule('UIAutomationCore.dll')
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\client\_generate.py", line 110, in GetModule
    mod = _CreateWrapper(tlib, pathname)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\client\_generate.py", line 172, in _CreateWrapper
    generate_module(tlib, ofi, pathname)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\tlbparser.py", line 750, in generate_module
    gen.generate_code(list(items.values()), filename=pathname)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 241, in generate_code
    self.generate_all(items)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 189, in generate_all
    self.generate(item)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 185, in generate
    mth(item)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 701, in ComInterface
    self.generate(itf.get_head())
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 185, in generate
    mth(item)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 722, in ComInterfaceHead
    self.generate(base.get_head())
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 185, in generate
    mth(item)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\tools\codegenerator.py", line 629, in External
    comtypes.client.GetModule(ext.tlib)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\client\_generate.py", line 146, in GetModule
    return _my_import("comtypes.gen." + modulename)
  File "C:\Program Files (x86)\Python36-32\lib\site-packages\comtypes\client\_generate.py", line 24, in _my_import
    return __import__(fullname, globals(), locals(), ['DUMMY'])
ModuleNotFoundError: No module named 'comtypes.gen.stdole'
```

**Comment by fxthomas** [comment]
Can confirm this on comtypes 1.1.7, Python 3.6.7 64-bits on Windows 10. I'm importing a big .tlb that generates about 6 modules (including `mscorlib` and `stdole`) in the `comtypes.gen` package.

This smells like a race condition, since :
* The error (after importing from a cleared cache) is often about `stdole`, but not always.
* Adding `time.sleep(0.1)` at the start of the `comtypes.client._generate._my_import` function makes everything work.

I tried adding `fsync` or `FlushFileBuffers` before `ofi.close` calls in the same module (assuming that imports failed because the file was still being written by the filesystem), but this did not work.

**Comment by nanonyme** [comment]
Note that it looks like you can possibly workaround this by
```
import comtypes.client
comtypes.client.gen_dir = None
```
to convince comtypes that generating files into disk is not safe. Documentation says in this case all modules are generated in memory.

**Comment by astrojuanlu** [comment]
I tried @nanonyme solution, but get this error:

```
In [1]: import comtypes.client
   ...: comtypes.client.gen_dir = None

In [2]: from win32api import GetSystemMetrics
   ...: import comtypes
   ...:
   ...: # Start new instance of STK
   ...: from comtypes.client import CreateObject
   ...: uiApplication = CreateObject('STK12.Application')
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-2-043297826ac5> in <module>
      4 # Start new instance of STK
      5 from comtypes.client import CreateObject
----> 6 uiApplication = CreateObject('STK12.Application')

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\__init__.py in CreateObject(progid, clsctx, machine, interface, dynamic, pServerInfo)
    248     if dynamic:
    249         return comtypes.client.dynamic.Dispatch(obj)
--> 250     return _manage(obj, clsid, interface=interface)
    251
    252 def CoGetObject(displayname, interface=None, dynamic=False):

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\__init__.py in _manage(obj, clsid, interface)
    186     obj.__dict__['__clsid'] = str(clsid)
    187     if interface is None:
--> 188         obj = GetBestInterface(obj)
    189     return obj
    190

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\__init__.py in GetBestInterface(punk)
    108
    109     # import the wrapper, generating it on demand
--> 110     mod = GetModule(tlib)
    111     # Python interface class
    112     interface = getattr(mod, itf_name)

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\_generate.py in GetModule(tlib)
    108
    109     # create and import the module
--> 110     mod = _CreateWrapper(tlib, pathname)
    111     try:
    112         modulename = tlib.GetDocumentation(-1)[0]

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\_generate.py in _CreateWrapper(tlib, pathname)
    170     # XXX use logging!
    171     logger.info("# Generating comtypes.gen.%s", modname)
--> 172     generate_module(tlib, ofi, pathname)
    173
    174     if comtypes.client.gen_dir is None:

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\tlbparser.py in generate_module(tlib, ofi, pathname)
    748                     )
    749
--> 750     gen.generate_code(list(items.values()), filename=pathname)
    751
    752 # -eof-

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\codegenerator.py in generate_code(self, items, filename)
    239             loops += 1
    240             self.more = set()
--> 241             self.generate_all(items)
    242
    243             items |= self.more

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\codegenerator.py in generate_all(self, items)
    187     def generate_all(self, items):
    188         for item in items:
--> 189             self.generate(item)
    190
    191     def _make_relative_path(self, path1, path2):

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\codegenerator.py in generate(self, item)
    183         # before actually generating the code.
    184         self.done.add(item)
--> 185         mth(item)
    186
    187     def generate_all(self, items):

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\codegenerator.py in CoClass(self, coclass)
    675
    676         for itf, idlflags in coclass.interfaces:
--> 677             self.generate(itf.get_head())
    678         implemented = []
    679         sources = []

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\codegenerator.py in generate(self, item)
    183         # before actually generating the code.
    184         self.done.add(item)
--> 185         mth(item)
    186
    187     def generate_all(self, items):

~\anaconda3\envs\stk\lib\site-packages\comtypes\tools\codegenerator.py in External(self, ext)
    627         self._externals[libdesc] = modname
    628         print("import", modname, file=self.imports)
--> 629         comtypes.client.GetModule(ext.tlib)
    630
    631     def Constant(self, tp):

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\_generate.py in GetModule(tlib)
    108
    109     # create and import the module
--> 110     mod = _CreateWrapper(tlib, pathname)
    111     try:
    112         modulename = tlib.GetDocumentation(-1)[0]

~\anaconda3\envs\stk\lib\site-packages\comtypes\client\_generate.py in _CreateWrapper(tlib, pathname)
    175         code = ofi.getvalue()
    176         mod = types.ModuleType(fullname)
--> 177         mod.__file__ = os.path.join(os.path.abspath(comtypes.gen.__path__[0]),
    178                                     "<memory>")
    179         exec(code, mod.__dict__)

TypeError: '_NamespacePath' object is not subscriptable
```

But I guess #172 fixes the issue?

**Comment by nanonyme** [comment]
It should once released. Note if doing what I suggested results in errors, that sounds like another bug. Documentation says this library supports install as zip where there is not disk and in that case modules are generated in memory. It worked when I tested it but it was a very simple case.

**Comment by astrojuanlu** [comment]
I manually patched the file as done in #172 and the problem is gone, at least :muscle:

**Comment by vasily-v-ryabov** [comment]
Guys (@astrojuanlu @nanonyme ) I'm planning to publish a new release in October.

@nanonyme special thanks for your great help!

**Comment by mohammed-workfellow** [comment]
I had similar issue recently (Windows 11 - Python 3.10.4 - comtypes 1.1.11). Comtypes was missing some resources. I tracked the issue, I found that `Lib/comtypes/gen/` where many resources were missing, `stdole` file wasn't there and one of the long named files was  semi-empty. The solution for me was two steps:
1. Follow the advice in https://github.com/enthought/comtypes/issues/161#issuecomment-689787596 comment by setting:
```
import comtypes.client
comtypes.client.gen_dir = None 
```

2. Clear cache using @vasily-v-ryabov advice from a different issue by calling `.venv/Scripts/clear_comtypes_cache.py`

I had all sort of issues and this procedure helped me a lot. I am writing this for developers debugging comtypes issues in the future.
